<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 4 Mapping Bulk data | neuRoDev</title>
<meta name="author" content="Erik Bot &amp; Asia Zonca">
<meta name="description" content="4.1 Quickstart These are the main functions to map a bulk RNA-sequencing dataset on the corticogenesis resource. This implies that the data are already normalized. Below we report a detailed...">
<meta name="generator" content="bookdown 0.43 with bs4_book()">
<meta property="og:title" content="Chapter 4 Mapping Bulk data | neuRoDev">
<meta property="og:type" content="book">
<meta property="og:description" content="4.1 Quickstart These are the main functions to map a bulk RNA-sequencing dataset on the corticogenesis resource. This implies that the data are already normalized. Below we report a detailed...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 4 Mapping Bulk data | neuRoDev">
<meta name="twitter:description" content="4.1 Quickstart These are the main functions to map a bulk RNA-sequencing dataset on the corticogenesis resource. This implies that the data are already normalized. Below we report a detailed...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.9.0/transition.js"></script><script src="libs/bs3compat-0.9.0/tabs.js"></script><script src="libs/bs3compat-0.9.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><link href="libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="libs/htmlwidgets-1.6.4/htmlwidgets.js"></script><link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="libs/datatables-binding-0.33/datatables.js"></script><link href="libs/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="libs/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="libs/dt-core-1.13.6/js/jquery.dataTables.min.js"></script><link href="libs/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet">
<script src="libs/crosstalk-1.2.1/js/crosstalk.min.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="libs/style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">neuRoDev</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> Summary</a></li>
<li><a class="" href="network.html"><span class="header-section-number">2</span> Network exploration</a></li>
<li><a class="" href="analysis.html"><span class="header-section-number">3</span> Analysis Tools</a></li>
<li><a class="active" href="mapping-bulk.html"><span class="header-section-number">4</span> Mapping Bulk data</a></li>
<li><a class="" href="mapping-sc.html"><span class="header-section-number">5</span> Mapping Single-cell data</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/davilavelderrainlab/neuRoDev/blob/main/book/">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="mapping-bulk" class="section level1" number="4">
<h1>
<span class="header-section-number">4</span> Mapping Bulk data<a class="anchor" aria-label="anchor" href="#mapping-bulk"><i class="fas fa-link"></i></a>
</h1>
<div id="quickstart-1" class="section level2" number="4.1">
<h2>
<span class="header-section-number">4.1</span> Quickstart<a class="anchor" aria-label="anchor" href="#quickstart-1"><i class="fas fa-link"></i></a>
</h2>
<p>These are the main functions to map a bulk RNA-sequencing dataset on the corticogenesis resource. This implies that the data are already normalized. Below we report a detailed description and examples on how to perform mapping and further related analyses.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#STEP 1: load libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">neuRoDev</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/r-gregmisc/gtools">gtools</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">#STEP 2: load data </span></span>
<span><span class="va">corticogenesis_sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/neuRoDev/man/corticogenesis.sce.html">corticogenesis.sce</a></span><span class="op">(</span>directory <span class="op">=</span> <span class="st">'~/Downloads'</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">'my_data'</span><span class="op">)</span> <span class="co">#choose your data object</span></span>
<span></span>
<span><span class="co">#STEP 3: mapping</span></span>
<span><span class="va">mapped_bulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/neuRoDev/man/mapNetwork.html">mapNetwork</a></span><span class="op">(</span>net <span class="op">=</span> <span class="va">corticogenesis_sce</span>, </span>
<span>                               new_profiles <span class="op">=</span> <span class="va">data</span><span class="op">)</span></span>
<span><span class="co">#STEP 4: inspect mapping</span></span>
<span><span class="va">mapped_bulk</span><span class="op">$</span><span class="va">new_plot</span></span>
<span><span class="va">mapped_bulk</span><span class="op">$</span><span class="va">annotation</span><span class="op">$</span><span class="va">Barplot</span></span>
<span><span class="va">mapped_bulk</span><span class="op">$</span><span class="va">annotation</span><span class="op">$</span><span class="va">Best.Annotation</span></span></code></pre></div>
</div>
<div id="loading-necessary-libraries-and-data-2" class="section level2" number="4.2">
<h2>
<span class="header-section-number">4.2</span> Loading necessary libraries and data<a class="anchor" aria-label="anchor" href="#loading-necessary-libraries-and-data-2"><i class="fas fa-link"></i></a>
</h2>
<div id="loading-the-right-libraries-2" class="section level3" number="4.2.1">
<h3>
<span class="header-section-number">4.2.1</span> Loading the right libraries<a class="anchor" aria-label="anchor" href="#loading-the-right-libraries-2"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># functions package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">neuRoDev</span><span class="op">)</span></span>
<span><span class="co"># extra packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/ComplexHeatmap">ComplexHeatmap</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/r-gregmisc/gtools">gtools</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/">edgeR</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div id="loading-the-networks-necessary-objects-2" class="section level3" number="4.2.2">
<h3>
<span class="header-section-number">4.2.2</span> Loading the networks necessary objects<a class="anchor" aria-label="anchor" href="#loading-the-networks-necessary-objects-2"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">corticogenesis_sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/neuRoDev/man/corticogenesis.sce.html">corticogenesis.sce</a></span><span class="op">(</span>directory <span class="op">=</span> <span class="st">'~/Downloads'</span><span class="op">)</span></span>
<span><span class="va">neurogenesis_sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/neuRoDev/man/neurogenesis.sce.html">neurogenesis.sce</a></span><span class="op">(</span>directory <span class="op">=</span> <span class="st">'~/Downloads'</span><span class="op">)</span></span>
<span><span class="va">gliogenesis_sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/neuRoDev/man/gliogenesis.sce.html">gliogenesis.sce</a></span><span class="op">(</span>directory <span class="op">=</span> <span class="st">'~/Downloads'</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div id="mapping-a-bulk-cell-rnaseq-dataset" class="section level2" number="4.3">
<h2>
<span class="header-section-number">4.3</span> Mapping a bulk cell RNAseq dataset<a class="anchor" aria-label="anchor" href="#mapping-a-bulk-cell-rnaseq-dataset"><i class="fas fa-link"></i></a>
</h2>
<p>This example dataset from <a href="https://www.nature.com/articles/s41593-021-00802-y">Gordon et al. (2021)</a> can be downloaded from: …
This dataset contains bulk RNAseq expression of organoid samples at different time points. For the scope of the analyses, we pre-processed the data, removing batch-effect using the package <code>sva</code>, and then computed Counts Per Million (CPM) (log-normalized, saved in <code>logcounts</code>). Pre-processing steps can be done differently and depending on your own dataset characteristics, but the following steps shown in this tutorial will apply to any bulk RNAseq samples. A lack of correct pre-processing steps may affect the results and interpretation of the following analyses.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bulk_sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">'~/Downloads/bulk_sce.rds'</span><span class="op">)</span></span></code></pre></div>
<p>We average between samples with the same differentiation day. If you are interested in variability between replicates, this step can be avoided.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bulk_average</span> <span class="op">&lt;-</span> <span class="fu">neuRoDev</span><span class="fu">:::</span><span class="fu"><a href="https://rdrr.io/pkg/neuRoDev/man/get_column_group_average.html">get_column_group_average</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">logcounts</a></span><span class="op">(</span><span class="va">bulk_sce</span><span class="op">)</span>, </span>
<span>                                                    <span class="va">bulk_sce</span><span class="op">$</span><span class="va">`differentiation day`</span><span class="op">)</span></span></code></pre></div>
<p>Differently than scRNAseq dataset mapping (see <a href="mapping-sc.html#mapping-sc">Chapter Mapping Single-cell data</a>), bulk RNAseq datasets don’t need further processing into transcriptional profiles, as they already represent a pool of cells sequenced together. Hence, bulk samples can be directly mapped using the <strong>mapNetwork</strong> function, that requires the following main inputs:</p>
<ul>
<li>
<code>net</code>: the reference network to use</li>
<li>
<code>new_profiles</code>: the transcriptomic profiles to map</li>
</ul>
<p>Other useful inputs can be checked by <strong>?mapNetwork</strong></p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mapped_bulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/neuRoDev/man/mapNetwork.html">mapNetwork</a></span><span class="op">(</span>net <span class="op">=</span> <span class="va">corticogenesis_sce</span>, </span>
<span>                          new_profiles <span class="op">=</span> <span class="va">bulk_average</span><span class="op">)</span></span></code></pre></div>
<p>We can now visualize the network with the mapped clusters on top:</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mapped_bulk</span><span class="op">$</span><span class="va">new_plot</span></span></code></pre></div>
<div class="figure">
<span style="display:block;" id="fig:ch5-fig1"></span>
<img src="libs/4.1-Mapping_bulkRNAseq_files/figure-html/ch5-fig1-1.png" alt="Corticogenesis network with mapped samples." width="672"><p class="caption">
Figure 4.1: Corticogenesis network with mapped samples.
</p>
</div>
<p>From the visualization, we can already appreciate how the samples at early days (d25, d50) map close to radial glia clusters, moving towards immature neurons in days 75, 100 and 150.
Nevertheless, from day 200 we can see that there is a unexpected transcriptional switch and samples return to map closer to radial glia. This could mean either that they somehow went back to a more undifferentiated state, or that they start looking more like mature glia (astrocytes in particular, which are similar to radial glia, also biologically). To investigate in deep these differences, it is possible to leverage the analyses tools described below.</p>
<p>Since newly mapped samples will locate close to the most similar reference clusters, it is possible to qualitatively understand the most similar subclass for each mapped point. To quantify this similarity, we developed a score per subclass per mapped sample, using the 15 most similar reference clusters’ annotations (the number of reference clusters to use can be set by changing <code>n_nearest</code> in the <strong>mapNetwork</strong> function). The score can be visualized by:</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mapped_bulk</span><span class="op">$</span><span class="va">annotation</span><span class="op">$</span><span class="va">Barplot</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<span style="display:block;" id="fig:ch5-fig2"></span>
<img src="libs/4.1-Mapping_bulkRNAseq_files/figure-html/ch5-fig2-1.png" alt="Annotation scores across samples" width="95%"><p class="caption">
Figure 4.2: Annotation scores across samples
</p>
</div>
<p>Days 250, 300 and 400 have a small resemblance to astrocytes, but this is lost in days 500 and 600.</p>
<p>To get a single label for each mapped cluster, we take the maximum of these scores.
This depicts the most similar subclass among the closest neighbors.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mapped_bulk</span><span class="op">$</span><span class="va">annotation</span><span class="op">$</span><span class="va">Best.Annotation</span></span>
<span><span class="co">#&gt;     d25     d50     d75    d100    d150    d200    d250 </span></span>
<span><span class="co">#&gt;    "RG"    "RG" "imNIT" "imNIT" "imNIT"    "RG"    "RG" </span></span>
<span><span class="co">#&gt;    d300    d400    d500    d600 </span></span>
<span><span class="co">#&gt;    "RG"    "RG"    "RG"    "RG"</span></span></code></pre></div>
<p>The same quantification can be done for stages instead of subclasses, either by specifying <code>color_attr = 'Stages'</code> in the <strong>mapNetwork</strong> function, or by running the <strong>annotateMapping</strong> function, which requires the following inputs:</p>
<ul>
<li>
<code>net</code>: the reference network to use.</li>
<li>
<code>new_cor</code>: the correlation between mapped points and reference clusters, can be found by <code>mapped_object$new_cor</code>.</li>
<li>
<code>color_attr</code>: the annotation label to use, in this case ‘Stages’.</li>
</ul>
<p>Our mapping strategy allows the mapping of any expression matrix, thus we added a quantification of the mapping confidence, described with two diverse measures:</p>
<ol style="list-style-type: lower-alpha">
<li><p><strong>local score</strong> (<code>mapped_object$annotation$Mapping.Confidence</code>) for each mapped point, showing average of the 15 highest correlation values that mapped point has with the reference clusters.</p></li>
<li><p><strong>global score</strong> (<code>mapped_object$annotation$Global.Confidence</code>), describing the average confidence among all mapped point.</p></li>
</ol>
<p>The confidence values can be accessed in the following way:</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># global confidence</span></span>
<span><span class="va">mapped_bulk</span><span class="op">$</span><span class="va">annotation</span><span class="op">$</span><span class="va">Global.Confidence</span></span>
<span><span class="co">#&gt; [1] 0.6783803</span></span>
<span></span>
<span><span class="co"># per cluster confidence</span></span>
<span><span class="va">mapped_bulk</span><span class="op">$</span><span class="va">annotation</span><span class="op">$</span><span class="va">Mapping.Confidence</span></span>
<span><span class="co">#&gt;       d25       d50       d75      d100      d150      d200 </span></span>
<span><span class="co">#&gt; 0.7468790 0.7425336 0.7363235 0.7106007 0.6891449 0.6729147 </span></span>
<span><span class="co">#&gt;      d250      d300      d400      d500      d600 </span></span>
<span><span class="co">#&gt; 0.6236117 0.6462351 0.6386074 0.6045623 0.6507705</span></span></code></pre></div>
<p>We can see that generally the mapped samples have relatively high confidence, but it decreases in the last days compared to the first.</p>
<p>A further analysis is mapping the samples directly on the eTraces, to see how similar they are to clusters that express more or less certain genes.
The function to perform such analysis is <strong>map_eTrace</strong>, which requires the following inputs:</p>
<ul>
<li>
<code>net</code>: the reference network to use.</li>
<li>
<code>mapped_obj</code>: the result of <strong>mapNetwork</strong>.</li>
<li>the inputs required to run <strong>plot_eTrace</strong> (see the <a href="analysis.html#analysis">Chapter Analysis tools</a>).</li>
</ul>
<p>We can visualize the expression of preferentially expressed genes in the relevant subclasses.
We report gene sets preferentially expressed in each reference subclass or subclass group here: …</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">corticogenesis_pe_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"~/Downloads/corticogenesis_subclass_preferential_genes.rds"</span><span class="op">)</span></span>
<span><span class="va">neurogenesis_pe_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"~/Downloads/neurogenesis_subclass_preferential_genes.rds"</span><span class="op">)</span></span>
<span><span class="va">gliogenesis_pe_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"~/Downloads/gliogenesis_subclass_preferential_genes.rds"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/neuRoDev/man/map_eTrace.html">map_eTrace</a></span><span class="op">(</span>net <span class="op">=</span> <span class="va">corticogenesis_sce</span>, </span>
<span>           mapped_obj <span class="op">=</span> <span class="va">mapped_bulk</span>, </span>
<span>           genes <span class="op">=</span> <span class="va">corticogenesis_pe_genes</span><span class="op">$</span><span class="va">RG</span>, </span>
<span>           main <span class="op">=</span> <span class="st">'RG'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/neuRoDev/man/map_eTrace.html">map_eTrace</a></span><span class="op">(</span>net <span class="op">=</span> <span class="va">corticogenesis_sce</span>, </span>
<span>           mapped_obj <span class="op">=</span> <span class="va">mapped_bulk</span>, </span>
<span>           genes <span class="op">=</span> <span class="va">corticogenesis_pe_genes</span><span class="op">$</span><span class="va">miEN.imNIT.eEN</span>, </span>
<span>           main <span class="op">=</span> <span class="st">'immEN'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/neuRoDev/man/map_eTrace.html">map_eTrace</a></span><span class="op">(</span>net <span class="op">=</span> <span class="va">corticogenesis_sce</span>, </span>
<span>           mapped_obj <span class="op">=</span> <span class="va">mapped_bulk</span>, </span>
<span>           genes <span class="op">=</span> <span class="va">corticogenesis_pe_genes</span><span class="op">$</span><span class="va">L6IT.L6NIT.L35IT.L23IT.SST.VIP.PVALB.LAMP5</span>, </span>
<span>           main <span class="op">=</span> <span class="st">'matNEU'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/neuRoDev/man/map_eTrace.html">map_eTrace</a></span><span class="op">(</span>net <span class="op">=</span> <span class="va">corticogenesis_sce</span>, </span>
<span>           mapped_obj <span class="op">=</span> <span class="va">mapped_bulk</span>, </span>
<span>           genes <span class="op">=</span> <span class="va">corticogenesis_pe_genes</span><span class="op">$</span><span class="va">Astro</span>, </span>
<span>           main <span class="op">=</span> <span class="st">'Astrocytes'</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<span style="display:block;" id="fig:ch5-fig3"></span>
<img src="libs/4.1-Mapping_bulkRNAseq_files/figure-html/ch5-fig3-1.png" alt="Preferential genes eTrace in corticogenesis and mapped points." width="50%"><img src="libs/4.1-Mapping_bulkRNAseq_files/figure-html/ch5-fig3-2.png" alt="Preferential genes eTrace in corticogenesis and mapped points." width="50%"><img src="libs/4.1-Mapping_bulkRNAseq_files/figure-html/ch5-fig3-3.png" alt="Preferential genes eTrace in corticogenesis and mapped points." width="50%"><img src="libs/4.1-Mapping_bulkRNAseq_files/figure-html/ch5-fig3-4.png" alt="Preferential genes eTrace in corticogenesis and mapped points." width="50%"><p class="caption">
Figure 4.3: Preferential genes eTrace in corticogenesis and mapped points.
</p>
</div>
<p>The eTrace of radial glia markers describes the same pattern shown when looking at the mapping on the network: the mapped points form a circle from early time points to late ones, that goes back to showing resemblance to early radial glia clusters. This further indicates that not only the look more similar to radial glia than neurons, but also that they look more similar to early radial glia compared to late ones.
By looking instead at the eTrace of immature excitatory neuron markers, we can see that samples from d75 to d200 are above the average line (d200 being the lowest), pointing to the fact that only these time points show a similarity to immature neurons. By plotting instead the eTrace of mature neurons, no sample shows particularly high values, suggesting that organoids do not recapitulate the full maturity described by in vivo samples. Lastly, the eTrace of astrocyte markers shows that days from 250 to 400 are indeed higher than the other samples (but also late radial glia clusters show relatively high values), but clearly that days 500 and 600 are much lower.</p>
<p>We show here a function (<strong>plotTrends</strong>) to visualize the expression of preferentially expressed genes in the bulk samples, as an example of further possible analyses that can be performed with the provided resources. The function is not included in the main functions of the package, but can be accessed by <code><a href="https://rdrr.io/pkg/neuRoDev/man/plotTrends.html">neuRoDev::plotTrends</a></code>. The function requires the following inputs:</p>
<ul>
<li>
<code>net</code>: the reference network to use.</li>
<li>
<code>pref_exp_genes</code>: the preferentially expressed genes list.</li>
<li>
<code>sce</code>: the external object in the form of a SingleCellExperiment object.</li>
<li>
<code>profiles</code>: the actual expression matrix used for the mapping.</li>
<li>
<code>coldata</code>: where to find the information of the replicates used for averaging.</li>
<li>
<code>subclass</code>: if indicated as a vector it selects only a subset of reference subclasses. Defaults to all subclasses.</li>
<li>
<code>ylim</code>: the lower and upper limits of the y-axis (defaults to NULL, which goes back to the automatic R plot ylim).</li>
<li>
<code>together</code>: defines if the different preferentially expressed genes should be plotted together (adjust <code>ylim</code> accordingly) (defaults to FALSE).</li>
<li>
<code>relative</code>: if TRUE, the expression is scaled from 0 to 1 (defaults to FALSE).</li>
</ul>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">neuRoDev</span><span class="fu">:::</span><span class="fu"><a href="https://rdrr.io/pkg/neuRoDev/man/plotTrends.html">plotTrends</a></span><span class="op">(</span>net <span class="op">=</span> <span class="va">corticogenesis_sce</span>, </span>
<span>           pref_exp_genes <span class="op">=</span> <span class="va">corticogenesis_pe_genes</span>, </span>
<span>           subclass <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Astro'</span>, <span class="st">'cRG'</span><span class="op">)</span>, </span>
<span>           sce <span class="op">=</span> <span class="va">bulk_sce</span>, </span>
<span>           profiles <span class="op">=</span> <span class="va">bulk_average</span>, </span>
<span>           coldata <span class="op">=</span> <span class="st">'differentiation day'</span>, </span>
<span>           together <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>           ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2.5</span>, <span class="fl">5.5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<span style="display:block;" id="fig:ch5-fig4"></span>
<img src="libs/4.1-Mapping_bulkRNAseq_files/figure-html/ch5-fig4-1.png" alt="Preferential genes of astrocytes and cycling radial glia expression in mapped data points." width="95%"><p class="caption">
Figure 4.4: Preferential genes of astrocytes and cycling radial glia expression in mapped data points.
</p>
</div>
<p>Astrocyte markers increase up to day 400, but then decrease again and are overcome by cycling radial glia markers at time points (d500, d600).</p>
<p>This is consistent with the expression levels of GO processes manually curated in our resource. These are available at: … and represent curated pathways relevant in corticogenesis and neurogenesis.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">corticogenesis_GO</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">'~/Downloads/corticogenesis_GOBP_genesets.rds'</span><span class="op">)</span></span>
<span><span class="va">neurogenesis_GO</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">'~/Downloads/neurogenesis_GOBP_genesets.rds'</span><span class="op">)</span></span></code></pre></div>
<p>To assess the expression levels of these processes, we first have to obtain the average expression of those gene sets in the mapped data:</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bulk_cortico_GOBP</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">corticogenesis_GO</span>, <span class="kw">function</span><span class="op">(</span><span class="va">g</span><span class="op">)</span> <span class="op">{</span></span>
<span>  </span>
<span>  <span class="va">g</span> <span class="op">&lt;-</span> <span class="va">g</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">g</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">bulk_sce</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span>  </span>
<span>  <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html">colMeans</a></span><span class="op">(</span><span class="va">bulk_average</span><span class="op">[</span><span class="va">g</span>,<span class="op">]</span><span class="op">)</span></span>
<span>  </span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">bulk_cortico_GOBP</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">" \\(GO:[0-9]+\\)$"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">bulk_cortico_GOBP</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">bulk_cortico_GOBP</span> <span class="op">&lt;-</span> <span class="va">bulk_cortico_GOBP</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/pkg/gtools/man/mixedsort.html">mixedorder</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">bulk_cortico_GOBP</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>Then, we can visualize the relative expression in an heatmap.
To compare this expression with the preferential expression of these genesets in our resource, it is possible to download the preferential expression of Gene Ontology Biological Processes (BP), Molecular Functions (MF), and Cellular Components (CC) here. Each object is a list containing preferential expression scores in one of the three reference networks in the three ontologies (BP, MF, CC). Each element of each list contains the activity (<code>activity</code>) derived from Gene Set Variation Analysis (one value per gene set in each cluster) and the preferential expression scores (<code>preferential</code>; one value per gene set in each subclass).</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select the path in which you have saved the objects!</span></span>
<span><span class="va">corticogenesis_preferential_GO</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">'~/Downloads/corticogenesis_preferential_GO.rds'</span><span class="op">)</span></span>
<span><span class="va">neurogenesis_preferential_GO</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">'~/Downloads/neurogenesis_preferential_GO.rds'</span><span class="op">)</span></span>
<span><span class="va">gliogenesis_preferential_GO</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">'~/Downloads/gliogenesis_preferential_GO.rds'</span><span class="op">)</span></span></code></pre></div>
<details><summary><strong>Show the code (simplifying labels)</strong>
</summary><div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cortico_subclass_groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">corticogenesis_preferential_GO</span><span class="op">$</span><span class="va">GO_Biological_Process_2025</span><span class="op">$</span><span class="va">preferential</span><span class="op">)</span></span>
<span><span class="va">cortico_subclass_groups</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">cortico_subclass_groups</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'L23IT'</span>, <span class="st">'L35IT'</span>, <span class="st">'L6IT'</span>, <span class="st">'L6NIT'</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'matEN'</span></span>
<span><span class="va">cortico_subclass_groups</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">cortico_subclass_groups</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'LAMP5'</span>, <span class="st">'SST'</span>, <span class="st">'PVALB'</span>, <span class="st">'VIP'</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'matIN'</span></span>
<span><span class="va">cortico_subclass_groups</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">cortico_subclass_groups</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'imIT'</span>, <span class="st">'imNIT'</span>, <span class="st">'eEN'</span>, <span class="st">'miEN'</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'immEN'</span></span>
<span><span class="va">cortico_subclass_groups</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">cortico_subclass_groups</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'eIN'</span>, <span class="st">'imIN'</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'immIN'</span></span>
<span><span class="va">cortico_subclass_groups</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">cortico_subclass_groups</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Opc'</span>, <span class="st">'Oligo'</span>, <span class="st">'Astro'</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'maGlia'</span></span>
<span><span class="va">cortico_subclass_groups</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">cortico_subclass_groups</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Micro'</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'miGlia'</span></span>
<span><span class="va">cortico_subclass_groups</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">cortico_subclass_groups</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'RG'</span>, <span class="st">'cRG'</span>, <span class="st">'IPC'</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'Prog'</span></span>
<span></span>
<span><span class="va">cortico_subclass_groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">cortico_subclass_groups</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Prog'</span>, <span class="st">'immEN'</span>, <span class="st">'immIN'</span>, <span class="st">'matEN'</span>, <span class="st">'matIN'</span>, <span class="st">'maGlia'</span>, <span class="st">'miGlia'</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cortico_gobp_groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'Astro'</span>, <span class="fl">4</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'Oligo'</span>, <span class="fl">3</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'Opc'</span>, <span class="fl">3</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'OligoGlia'</span>, <span class="fl">2</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'matEN'</span>, <span class="fl">4</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'matIN'</span>, <span class="fl">4</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'immEN'</span>, <span class="fl">4</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'Micro'</span>, <span class="fl">4</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'Prog'</span>, <span class="fl">4</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">'Stress'</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Prog'</span>, <span class="st">'immEN'</span>, <span class="st">'matEN'</span>, <span class="st">'matIN'</span>, <span class="st">'Opc'</span>, <span class="st">'Oligo'</span>, <span class="st">'OligoGlia'</span>, <span class="st">'Astro'</span>, <span class="st">'Micro'</span>, <span class="st">'Stress'</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cortico_gobp_pe</span> <span class="op">&lt;-</span> <span class="va">corticogenesis_preferential_GO</span><span class="op">$</span><span class="va">GO_Biological_Process_2025</span><span class="op">$</span><span class="va">preferential</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">corticogenesis_GO</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">cortico_gobp_pe</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">" \\(GO:[0-9]+\\)$"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">cortico_gobp_pe</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</details><details><summary><strong>Show the code (plotting heatmap)</strong>
</summary><div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ht1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">cortico_gobp_pe</span>,</span>
<span>  name <span class="op">=</span> <span class="st">'preferential\nexpression'</span>,</span>
<span>  width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">cortico_gobp_pe</span><span class="op">)</span> <span class="op">*</span> <span class="fl">4.2</span>, <span class="st">'mm'</span><span class="op">)</span>,</span>
<span>  height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">cortico_gobp_pe</span><span class="op">)</span> <span class="op">*</span> <span class="fl">4.2</span>, <span class="st">'mm'</span><span class="op">)</span>,</span>
<span>  show_row_dend <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  show_column_dend <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  row_title_rot <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  column_title_rot <span class="op">=</span> <span class="fl">90</span>,</span>
<span>  column_split <span class="op">=</span> <span class="va">cortico_subclass_groups</span>,</span>
<span>  cluster_column_slices <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  row_split <span class="op">=</span> <span class="va">cortico_gobp_groups</span>,</span>
<span>  cluster_row_slices <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  column_title_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">ht2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">bulk_cortico_GOBP</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  name <span class="op">=</span> <span class="st">'scaled\nexpression'</span>,</span>
<span>  width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">bulk_cortico_GOBP</span><span class="op">)</span> <span class="op">*</span> <span class="fl">4.2</span>, <span class="st">'mm'</span><span class="op">)</span>,</span>
<span>  cluster_columns <span class="op">=</span> <span class="cn">FALSE</span>, right_annotation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/rowAnnotation.html">rowAnnotation</a></span><span class="op">(</span>max <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/anno_barplot.html">anno_barplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">bulk_cortico_GOBP</span>, <span class="fl">1</span>, <span class="va">max</span><span class="op">)</span>, gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">'black'</span><span class="op">)</span><span class="op">)</span>, annotation_name_side <span class="op">=</span> <span class="st">'top'</span>, annotation_name_rot <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">ht_list</span> <span class="op">&lt;-</span> <span class="va">ht1</span> <span class="op">+</span> <span class="va">ht2</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html">draw</a></span><span class="op">(</span><span class="va">ht_list</span>, heatmap_legend_side <span class="op">=</span> <span class="st">"left"</span><span class="op">)</span></span></code></pre></div>
</details><div class="figure" style="text-align: left">
<span style="display:block;" id="fig:ch5-fig5"></span>
<img src="libs/4.1-Mapping_bulkRNAseq_files/figure-html/ch5-fig5-1.png" alt="Heatmap of corticogenesis-specific GO biological processes in resource and mapped samples." width="98%"><p class="caption">
Figure 4.5: Heatmap of corticogenesis-specific GO biological processes in resource and mapped samples.
</p>
</div>
<p>From this heatmap we can see that radial glia biological processes are highly expressed in the first 2-3 time points, but also in the last two. This, together with a high expression of stress-related processes (which increases as time progresses) suggests that indeed the last days are more stressed rather than being more similar to astrocytes. Also neuronal processes start around day 75, but decrease around days 300-400.</p>
<p>The same type of analysis can be done with the neurogenesis network:</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bulk_neuro_GOBP</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">neurogenesis_GO</span>, <span class="kw">function</span><span class="op">(</span><span class="va">g</span><span class="op">)</span> <span class="op">{</span></span>
<span>  </span>
<span>  <span class="va">g</span> <span class="op">&lt;-</span> <span class="va">g</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">g</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">bulk_average</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span>  </span>
<span>  <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html">colMeans</a></span><span class="op">(</span><span class="va">bulk_average</span><span class="op">[</span><span class="va">g</span>,<span class="op">]</span><span class="op">)</span></span>
<span>  </span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">bulk_neuro_GOBP</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">" \\(GO:[0-9]+\\)$"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">bulk_neuro_GOBP</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">bulk_neuro_GOBP</span> <span class="op">&lt;-</span> <span class="va">bulk_neuro_GOBP</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/pkg/gtools/man/mixedsort.html">mixedorder</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">bulk_neuro_GOBP</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<details><summary><strong>Show the code (simplifying labels and plotting heatmap)</strong>
</summary><div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">neuro_gobp_pe</span> <span class="op">&lt;-</span> <span class="va">neurogenesis_preferential_GO</span><span class="op">$</span><span class="va">GO_Biological_Process_2025</span><span class="op">$</span><span class="va">activity</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">neurogenesis_GO</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="va">neuroTrends</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="va">cbind</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">neuro_gobp_pe</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/stats/smooth.spline.html">smooth.spline</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">neuro_gobp_pe</span><span class="op">[</span><span class="va">i</span>,<span class="op">]</span><span class="op">)</span>, <span class="va">neuro_gobp_pe</span><span class="op">[</span><span class="va">i</span>,<span class="op">]</span>, spar <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">$</span><span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">neuroTrends</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">neuro_gobp_pe</span><span class="op">)</span></span>
<span></span>
<span><span class="va">neuroTrends</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">neuroTrends</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">neurogenesis_GO</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">neuroTrends</span> <span class="op">&lt;-</span> <span class="va">neuroTrends</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">neuroTrends</span>, <span class="fl">1</span>, <span class="va">which.max</span><span class="op">)</span>, <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html">rowMeans</a></span><span class="op">(</span><span class="va">neuroTrends</span><span class="op">)</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">neuroTrends</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">" \\(GO:[0-9]+\\)$"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">neuroTrends</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">stage_palette</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">neurogenesis_sce</span><span class="op">$</span><span class="va">Stages_color</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">stage_palette</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">neurogenesis_sce</span><span class="op">$</span><span class="va">Stages</span><span class="op">)</span></span>
<span></span>
<span><span class="va">h1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">neuroTrends</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, </span>
<span>              cluster_columns <span class="op">=</span> <span class="cn">F</span>, </span>
<span>              cluster_rows <span class="op">=</span> <span class="cn">F</span>, </span>
<span>              bottom_annotation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html">HeatmapAnnotation</a></span><span class="op">(</span>df<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>stage<span class="op">=</span><span class="va">neurogenesis_sce</span><span class="op">$</span><span class="va">Stages</span><span class="op">)</span>, </span>
<span>                                                    col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>stage<span class="op">=</span><span class="va">stage_palette</span><span class="op">)</span>, </span>
<span>                                                    show_legend <span class="op">=</span> <span class="cn">F</span>, </span>
<span>                                                    show_annotation_name <span class="op">=</span> <span class="cn">F</span><span class="op">)</span>, </span>
<span>              name <span class="op">=</span> <span class="st">'inferred expression\n(z scaled)'</span>, </span>
<span>              height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">100</span>, <span class="st">'mm'</span><span class="op">)</span>, </span>
<span>              width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">30</span>, <span class="st">'mm'</span><span class="op">)</span>, </span>
<span>              row_names_max_width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">20</span>, <span class="st">"mm"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ht2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">bulk_neuro_GOBP</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">neuroTrends</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  name <span class="op">=</span> <span class="st">'scaled\nexpression'</span>,</span>
<span>  width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">bulk_neuro_GOBP</span><span class="op">)</span> <span class="op">*</span> <span class="fl">4.2</span>, <span class="st">'mm'</span><span class="op">)</span>,</span>
<span>  cluster_columns <span class="op">=</span> <span class="cn">FALSE</span>, right_annotation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/rowAnnotation.html">rowAnnotation</a></span><span class="op">(</span>max <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/anno_barplot.html">anno_barplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">bulk_neuro_GOBP</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">neuroTrends</span><span class="op">)</span>,<span class="op">]</span>, <span class="fl">1</span>, <span class="va">max</span><span class="op">)</span>, gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">'black'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, , row_names_max_width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">20</span>, <span class="st">"mm"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">ht_list</span> <span class="op">&lt;-</span> <span class="va">h1</span> <span class="op">+</span> <span class="va">ht2</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html">draw</a></span><span class="op">(</span><span class="va">ht_list</span>, </span>
<span>     heatmap_legend_side <span class="op">=</span> <span class="st">"left"</span><span class="op">)</span></span></code></pre></div>
</details><div class="figure">
<span style="display:block;" id="fig:ch5-fig6"></span>
<img src="libs/4.1-Mapping_bulkRNAseq_files/figure-html/ch5-fig6-1.png" alt="Heatmap of neuroogenesis-specific GO biological processes in resource and mapped samples." width="90%"><p class="caption">
Figure 4.6: Heatmap of neuroogenesis-specific GO biological processes in resource and mapped samples.
</p>
</div>
<p>We can see that neuroinflammatory response regulation increases over time, as well as regulation of neuron apoptotic process.</p>
<p>Another approach to confirm our findings is to obtain markers for each sample using the <code>edgeR</code> and <code>limma</code> packages and plot those genes in the eTraces.</p>
<p>We can obtain an edgeR object in the following way:</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dge_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html">DGEList</a></span><span class="op">(</span>counts<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html">counts</a></span><span class="op">(</span><span class="va">bulk_sce</span><span class="op">)</span>, </span>
<span>                   samples<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">bulk_sce</span><span class="op">)</span>, </span>
<span>                                 <span class="va">bulk_sce</span><span class="op">$</span><span class="va">`differentiation day`</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html">filterByExpr</a></span><span class="op">(</span><span class="va">dge_obj</span>, </span>
<span>                     group<span class="op">=</span><span class="va">bulk_sce</span><span class="op">$</span><span class="va">`differentiation day`</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dge_obj</span> <span class="op">&lt;-</span> <span class="va">dge_obj</span><span class="op">[</span><span class="va">keep</span>, , keep.lib.sizes<span class="op">=</span><span class="cn">FALSE</span><span class="op">]</span></span>
<span></span>
<span><span class="va">dge_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html">calcNormFactors</a></span><span class="op">(</span><span class="va">dge_obj</span><span class="op">)</span></span></code></pre></div>
<p>We can then compute differential expression between differentiation days:</p>
<details><summary><strong>Show the code (differential expression)</strong>
</summary><div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">unique_groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">bulk_sce</span><span class="op">$</span><span class="va">`differentiation day`</span><span class="op">)</span></span>
<span><span class="va">de_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">g</span> <span class="kw">in</span> <span class="va">unique_groups</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">binary_group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">bulk_sce</span><span class="op">$</span><span class="va">`differentiation day`</span> <span class="op">==</span> <span class="va">g</span>, <span class="va">g</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Not"</span>, <span class="va">g</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">binary_group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">binary_group</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">sample_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="st">'binary_group'</span> <span class="op">=</span> <span class="va">binary_group</span>, </span>
<span>                            <span class="st">'groups'</span> <span class="op">=</span> <span class="va">bulk_sce</span><span class="op">$</span><span class="va">`differentiation day`</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">design</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">binary_group</span>, data<span class="op">=</span><span class="va">sample_info</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">dge_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html">estimateDisp</a></span><span class="op">(</span><span class="va">dge_obj</span>, </span>
<span>                          <span class="va">design</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFit.html">glmQLFit</a></span><span class="op">(</span><span class="va">dge_obj</span>, </span>
<span>                  <span class="va">design</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">qlf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html">glmQLFTest</a></span><span class="op">(</span><span class="va">fit</span>, </span>
<span>                    coef<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html">topTags</a></span><span class="op">(</span><span class="va">qlf</span>, </span>
<span>                 n<span class="op">=</span><span class="cn">Inf</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">res</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">table</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">table</span><span class="op">$</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span>,<span class="op">]</span></span>
<span>  </span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sign.html">sign</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/sign.html">sign</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">de_list</span><span class="op">[[</span><span class="va">g</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">res</span></span>
<span>    <span class="kw">next</span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="va">v1</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html">colMeans</a></span><span class="op">(</span><span class="va">bulk_average</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">]</span>,<span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">v2</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html">colMeans</a></span><span class="op">(</span><span class="va">bulk_average</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">logFC</span> <span class="op">&lt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">]</span>,<span class="op">]</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rank.html">rank</a></span><span class="op">(</span><span class="va">v1</span><span class="op">)</span><span class="op">[</span><span class="va">g</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/rank.html">rank</a></span><span class="op">(</span><span class="va">v2</span><span class="op">)</span><span class="op">[</span><span class="va">g</span><span class="op">]</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">res</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>,<span class="op">]</span>  </span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">res</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">logFC</span> <span class="op">&lt;</span> <span class="fl">0</span><span class="op">)</span>,<span class="op">]</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="va">de_list</span><span class="op">[[</span><span class="va">g</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">res</span></span>
<span><span class="op">}</span></span></code></pre></div>
</details>
We order the genes based on the (-log10) FDR, keeping only those that are also present in the corticogenesis resource.
<details><summary><strong>Show the code (filtering genes)</strong>
</summary><div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">de_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">de_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">i</span><span class="op">$</span><span class="va">score</span> <span class="op">&lt;-</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">i</span><span class="op">$</span><span class="va">FDR</span><span class="op">)</span></span>
<span>  <span class="va">i</span> <span class="op">&lt;-</span> <span class="va">i</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">i</span><span class="op">$</span><span class="va">score</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>,<span class="op">]</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">i</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">de_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">i</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">markers</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">i</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">i</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">corticogenesis_sce</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="va">markers</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">markers</span>, <span class="va">length</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
</details><div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">markers</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">g</span> <span class="op">&lt;-</span> <span class="va">markers</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">g</span> <span class="op">&lt;-</span> <span class="va">g</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">g</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">corticogenesis_sce</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/neuRoDev/man/plot_eTrace.html">plot_eTrace</a></span><span class="op">(</span>net <span class="op">=</span> <span class="va">corticogenesis_sce</span>, genes <span class="op">=</span> <span class="va">g</span>, main <span class="op">=</span> <span class="va">i</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="figure" style="text-align: left">
<span style="display:block;" id="fig:ch5-fig7"></span>
<img src="libs/4.1-Mapping_bulkRNAseq_files/figure-html/ch5-fig7-1.png" alt="eTrace of sample-specific differential genes in corticogenesis." width="48%"><img src="libs/4.1-Mapping_bulkRNAseq_files/figure-html/ch5-fig7-2.png" alt="eTrace of sample-specific differential genes in corticogenesis." width="48%"><img src="libs/4.1-Mapping_bulkRNAseq_files/figure-html/ch5-fig7-3.png" alt="eTrace of sample-specific differential genes in corticogenesis." width="48%"><img src="libs/4.1-Mapping_bulkRNAseq_files/figure-html/ch5-fig7-4.png" alt="eTrace of sample-specific differential genes in corticogenesis." width="48%"><img src="libs/4.1-Mapping_bulkRNAseq_files/figure-html/ch5-fig7-5.png" alt="eTrace of sample-specific differential genes in corticogenesis." width="48%"><img src="libs/4.1-Mapping_bulkRNAseq_files/figure-html/ch5-fig7-6.png" alt="eTrace of sample-specific differential genes in corticogenesis." width="48%"><img src="libs/4.1-Mapping_bulkRNAseq_files/figure-html/ch5-fig7-7.png" alt="eTrace of sample-specific differential genes in corticogenesis." width="48%"><img src="libs/4.1-Mapping_bulkRNAseq_files/figure-html/ch5-fig7-8.png" alt="eTrace of sample-specific differential genes in corticogenesis." width="48%"><img src="libs/4.1-Mapping_bulkRNAseq_files/figure-html/ch5-fig7-9.png" alt="eTrace of sample-specific differential genes in corticogenesis." width="48%"><img src="libs/4.1-Mapping_bulkRNAseq_files/figure-html/ch5-fig7-10.png" alt="eTrace of sample-specific differential genes in corticogenesis." width="48%"><p class="caption">
Figure 4.7: eTrace of sample-specific differential genes in corticogenesis.
</p>
</div>
<p>From this analysis we can appreciate how genes differentially expressed in the different samples are expressed in our reference. We can appreciate how genes differentially expressed in organoids from the early time points decrease their expression over time, particularly those of day 25.
From day 75 to day 200 we notice that the differentially expressed genes are mainly neuronal, and in particular for day 200, and they are highly expressed in immature neurons, decreasing along maturation. Interestingly, differentially expressed genes in days 300 and 400 are indeed more expressed by astrocytes and increase over time, while those specific for day 600 are higher in radial glia cells and microglia, which usually hints to a stress response (! microglia shouldn’t populate these organoids).</p>

</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="analysis.html"><span class="header-section-number">3</span> Analysis Tools</a></div>
<div class="next"><a href="mapping-sc.html"><span class="header-section-number">5</span> Mapping Single-cell data</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#mapping-bulk"><span class="header-section-number">4</span> Mapping Bulk data</a></li>
<li><a class="nav-link" href="#quickstart-1"><span class="header-section-number">4.1</span> Quickstart</a></li>
<li>
<a class="nav-link" href="#loading-necessary-libraries-and-data-2"><span class="header-section-number">4.2</span> Loading necessary libraries and data</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#loading-the-right-libraries-2"><span class="header-section-number">4.2.1</span> Loading the right libraries</a></li>
<li><a class="nav-link" href="#loading-the-networks-necessary-objects-2"><span class="header-section-number">4.2.2</span> Loading the networks necessary objects</a></li>
</ul>
</li>
<li><a class="nav-link" href="#mapping-a-bulk-cell-rnaseq-dataset"><span class="header-section-number">4.3</span> Mapping a bulk cell RNAseq dataset</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/davilavelderrainlab/neuRoDev/blob/main/book//blob/master/4.1-Mapping_bulkRNAseq.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/davilavelderrainlab/neuRoDev/blob/main/book//edit/master/4.1-Mapping_bulkRNAseq.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>neuRoDev</strong>" was written by Erik Bot &amp; Asia Zonca. It was last built on 2025-12-05.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
